// @applitools/eyes-testcafe/dist/api.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchClose = exports.Target = exports.CheckSettings = exports.Configuration = exports.Eyes = void 0;
const eyes_sdk_core_1 = require("@applitools/eyes-sdk-core");
const api = require("@applitools/eyes-api");
const legacy = require("./legacy");
const spec = require("./spec-driver");
process.env.APPLITOOLS_SCRIPT_RESULT_MAX_BYTE_LENGTH = '4718592';
process.env.APPLITOOLS_SCRIPT_REMOVE_REVERSE_PROXY_URL_PREFIXES = 'true';
const sdk = (0, eyes_sdk_core_1.makeSDK)({
    name: 'eyes.testcafe',
    version: require('../package.json').version,
    spec,
    VisualGridClient: require('@applitools/visual-grid-client'),
});
__exportStar(require("@applitools/eyes-api"), exports);
class Eyes extends legacy.LegacyTestCafeEyesMixin(api.Eyes) {
}
exports.Eyes = Eyes;
Eyes._spec = sdk;
class Configuration extends api.Configuration {
}
exports.Configuration = Configuration;
Configuration._spec = sdk;
class CheckSettings extends api.CheckSettings {
}
exports.CheckSettings = CheckSettings;
CheckSettings._spec = sdk;
exports.Target = CheckSettings;
class BatchClose extends api.BatchClose {
}
exports.BatchClose = BatchClose;
BatchClose._spec = sdk;
